{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SIVmac239M analysis pipeline \n",
    "### Modified from Athena Golfinos and Brandon Keele\n",
    "Last modified by Ryan V Moriarty, June 14, 2021"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Import modules \"\"\"\n",
    "import os\n",
    "import pandas as pd\n",
    "import glob\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import collections \n",
    "from fastqandfurious import fastqandfurious\n",
    "from fastqandfurious.fastqandfurious import entryfunc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" We also need to make sure that we have necessary information:\n",
    "    Where are the fastq files we need?\n",
    "    What experiment is this? \n",
    "    What index reads are there? \n",
    "    What barcodes are we looking for? \"\"\"\n",
    "\n",
    "ROOT = \"/Users/ryanmoriarty/Documents/SIV_barcoding_for_Pediatric_samples/\"\n",
    "EXPT = \"6346\"\n",
    "INDICES = pd.read_csv(ROOT + \"analysis_tools/index_ids.csv\", sep = ',', header = None)\n",
    "#idx = list(zip(INDICES[0], INDICES[1], INDICES[2]))\n",
    "BARCODES = [(b.id, b.seq) for b in SeqIO.parse(ROOT + \"/analysis_tools/SIV239MReferenceSequences.fasta\", \"fasta\")]\n",
    "P7 = \"TAGATCGTC\" #this is the P7 universal primer on the 5' end\n",
    "pre_index = \"GAGGTTCTGG\"\n",
    "pre_index_reverse_complement = \"CCAGAACCTC\"\n",
    "\n",
    "os.chdir(ROOT + EXPT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['6346_all_samples_R1.fastq']\n"
     ]
    }
   ],
   "source": [
    "# Make a list of the samples \n",
    "#print(\"Working directory:\", os.getcwd())\n",
    "samples = [f for f in glob.glob(\"*.fastq\")]\n",
    "sample_index = [s[s.find(\"P5.\"):s.find(\".fastq\")] for s in samples]\n",
    "zipped = [f for f in glob.glob(\"*.gz\")]\n",
    "fastqs = [ROOT+EXPT+\"/\" + file for file in glob.glob(\"*.fastq\")] \n",
    "fasta_seqs = [SeqIO.parse(open(f), \"fastq\") for f in fastqs]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fastq_file_lists(fastq_files):\n",
    "    # pull in the sequences \n",
    "    seqs = []\n",
    "    bufsize = 200000\n",
    "    with open(fastq_files, \"rb\") as fh:\n",
    "        it = fastqandfurious.readfastq_iter(fh, bufsize, entryfunc)\n",
    "        for sequence in it:\n",
    "            # sequence[1] is the nucleotide sequence, sequence[2] is the quality scores \n",
    "            seqs.append((str(sequence[1])[2:-2], str(sequence[2])[2:-2]))\n",
    "    \n",
    "    preidx = []\n",
    "    rcidx = []\n",
    "    preidx_not_found = []\n",
    "    for a in seqs:\n",
    "        if pre_index in a[0]:\n",
    "            preidx.append(a[0].index(pre_index))\n",
    "        elif pre_index_reverse_complement in a[0]: \n",
    "            rcidx.append(a[0].index(pre_index_reverse_complement))\n",
    "        else:\n",
    "            preidx_not_found.append(a)\n",
    "    seqs_rc = []\n",
    "    if len(rcidx) > len(preidx):\n",
    "        for a in seqs:\n",
    "             seqs_rc.append((str(Seq(a[0]).reverse_complement()), a[1][::-1]))\n",
    "        seqs = seqs_rc\n",
    "        print(\"samples are reverse transcribed \")\n",
    "            \n",
    "    return seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\" Sort the counter object to determine % barcodes \"\"\"\n",
    "\n",
    "def demultiplex(file, indices, write_files=True):\n",
    "    \"\"\"\n",
    "    Demultiplexes the Miseq file and separates them into individual samples\n",
    "    \n",
    "    Args:\n",
    "        file = the non-demultiplexed R1 or R2 file, save as .fastq\n",
    "        indices = the P5.XX used to label each sample\n",
    "        \n",
    "    Returns:\n",
    "        list of demultiplexed samples:\n",
    "            \"Unknown_tag\" - these are samples where the pre-index sequence \"GAGGTTCTGG\" or its reverese complement \n",
    "                was found, but the subseequent 8 base pairs, where the tag should be, was not present in the list of known \n",
    "                index tags. \n",
    "            \"Unindexed\" - these are samples where the pre-index sequence \"GAGGTTCTGG\" or its reverse complement was \n",
    "                NOT found, so we didn't look for the index tag either.\n",
    "            All other samples will be labeled with their index tag \"P5.XX\"\n",
    "                \n",
    "    \n",
    "    \"\"\"\n",
    "    \n",
    "    file_end = file[file.rfind(\"/\")+1:file.rfind(\".fastq\")]\n",
    "    all_seqs = fastq_file_lists(file)\n",
    "    \n",
    "    index_names = list(indices[0])\n",
    "    index_seqs = list(indices[1])\n",
    "    index_seqs_rc = list(indices[2])\n",
    "    \n",
    "    # find where the indices are hiding \n",
    "    dmux_samples = dict()\n",
    "    for a in all_seqs:\n",
    "        if pre_index in a[0]:\n",
    "            start = a[0].index(pre_index)+10\n",
    "            tag = a[0][start:start+8]\n",
    "            if tag in index_seqs:\n",
    "                name = index_names[index_seqs.index(tag)] + \"_rc\"\n",
    "                dmux_samples.setdefault(name, []).append(a)\n",
    "            elif tag in index_seqs_rc:\n",
    "                name = index_names[index_seqs_rc.index(tag)] \n",
    "                dmux_samples.setdefault(name, []).append(a)\n",
    "            else:\n",
    "                dmux_samples.setdefault(\"Unknown_tag\", []).append(a)\n",
    "        elif pre_index_reverse_complement in a[0]:\n",
    "            start = a[0].index(pre_index_reverse_complement)+10\n",
    "            tag = a[0][start:start+8]\n",
    "            if tag in index_seqs:\n",
    "                name = index_names[index_seqs.index(tag)] + \"_rc\"\n",
    "                dmux_samples.setdefault(name, []).append(a) \n",
    "            elif tag in index_seqs_rc:\n",
    "                name = index_names[index_seqs_rc.index(tag)] \n",
    "                dmux_samples.setdefault(name, []).append(a)\n",
    "            else:\n",
    "                dmux_samples.setdefault(\"Unknown_tag\", []).append(a)\n",
    "        else:\n",
    "            dmux_samples.setdefault(\"Unindexed\", []).append(a)\n",
    "    \n",
    "    index_keys = list(dmux_samples.keys())\n",
    "    real_tags_found = len(list(filter(lambda element: 'P' in element, index_keys)))\n",
    "    \n",
    "    if real_tags_found > 0:\n",
    "        if write_files:\n",
    "            for i in index_keys:\n",
    "                print(\"Writing sample index: \", i)\n",
    "                data = dmux_samples.get(i)\n",
    "                sample_fastq = open(file_end + \"_\" + i + \"_demultiplexed.fastq\", \"w\")\n",
    "                l = 0\n",
    "                for line in data:\n",
    "                    l += 1\n",
    "                    sample_fastq.write(\"@\" + i + \".\" + str(l) + \"\\n\" + line[0] + \"\\n+\\n\" + line[1] + \"\\n\" )\n",
    "                sample_fastq.close()\n",
    "    else:\n",
    "        print(\"No known index tags found, try different read.\")\n",
    "    \n",
    "    return dmux_samples"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
